src/libstd/sys/common/io.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use prelude::v1::*;
+use io;
+use io::ErrorKind;
+use io::Read;
+use slice::from_raw_parts_mut;
+
+pub const DEFAULT_BUF_SIZE: usize = 64 * 1024;
+
+pub fn read_to_end_zeroed<R: Read + ?Sized>(mut r: &mut R, buf: &mut Vec<u8>) -> io::Result<usize> {
+    unsafe { read_to_end(&mut r, buf, true) }
+}
+
+// This function is unsafe because it can optionally call the underlying
+// read function with a slice into uninitialized memory. The default
+// implementation of read_to_end for readers will zero out new memory in
+// the buf before passing it to read, but avoiding this zero can often
+// lead to a fairly significant perfomance win.
+//
+// Implementations using this method have to adhere to two guarantees:
+//  *  The implementation of read never reads the buffer provided.
+//  *  The implementation of read correctly reports how many bytes were written.
+pub unsafe fn read_to_end_uninitialized<R: Read + ?Sized>(mut r: &mut R, buf: &mut Vec<u8>)
+                                                          -> io::Result<usize> {
+    read_to_end(&mut r, buf, false)
+}
+
+// Used to fill a Vec from a Read type until there's no more data to come.
+// Uses an adaptive system to extend the vector when it fills. We want to
+// avoid paying to allocate and zero a huge chunk of memory if the reader only
+// has 4 bytes while still making large reads if the reader does have a ton
+// of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
+// time is 4,500 times (!) slower than this if the reader has a very small
+// amount of data to return.
+unsafe fn read_to_end(r: &mut Read, buf: &mut Vec<u8>, zero_buf: bool)
+                      -> io::Result<usize> {
+    let start_len = buf.len();
+    let mut len = start_len;
+    let mut new_write_size = 16;
+    let ret;
+    loop {
+        if len == buf.len() {
```

Comment:
Complexity rising, sorry about that that.

The code now is slightly different from what it was before, since `buf.len()` used to be the “capacity” level we aspired to read to. In the current code (uninitialized case), `buf.len()` is simply the amount we have read. So every iteration should hit this `len == buf.len()` case. Worst part is probably that on the last read, which was presumably short, we don't want to grow & read again, we just want to read again (and receive 0), so we'd save ourselves that 64k extra allocation in the vector.

Not sure if it matters, but it would be nice to write this loop in a different style. For example, it's not immediately obvious from the code that `buf_slice` is always inside the capacity of the vector.